### PR TITLE
coerce immutable arrays to normal arrays in the yum_package resource

### LINKED ISSUE #5812: Node attributes are read-only on package install with 12.19
I'm seeing the following failure in the tftp cookbook with 12.19 in Travis that doesn't occur locally with 12.18.

```
Recipe: tftp::server
  * yum_package[tftp-server] action install
    
    ================================================================================
    Error executing action `install` on resource 'yum_package[tftp-server]'
    ================================================================================
    
    Chef::Exceptions::ImmutableAttributeModification
    ------------------------------------------------
    Node attributes are read-only when you do not specify which precedence level to set. To set an attribute use code like `node.default["key"] = "value"'
    
    Resource Declaration:
    ---------------------
    # In /opt/kitchen/cache/cookbooks/tftp/recipes/server.rb
    
     26: package node['tftp']['pkgs']
     27: 
    
    Compiled Resource:
    ------------------
    # Declared in /opt/kitchen/cache/cookbooks/tftp/recipes/server.rb:26:in `from_file'
    
    yum_package("tftp-server") do
      package_name ["tftp-server"]
      action [:install]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      declared_type :package
      cookbook_name "tftp"
      recipe_name "server"
      flush_cache {:before=>false, :after=>false}
    end
    
    Platform:
    ---------
    x86_64-linux
    
Running handlers:
[2017-02-12T17:24:30+00:00] ERROR: Running exception handlers
Running handlers complete
[2017-02-12T17:24:30+00:00] ERROR: Exception handlers complete
Chef Client failed. 0 resources updated in 12 seconds
[2017-02-12T17:24:30+00:00] FATAL: Stacktrace dumped to /opt/kitchen/cache/chef-stacktrace.out
[2017-02-12T17:24:30+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2017-02-12T17:24:30+00:00] ERROR: yum_package[tftp-server] (tftp::server line 26) had an error: Chef::Exceptions::ImmutableAttributeModification: Node attributes are read-only when you do not specify which precedence level to set. To set an attribute use code like `node.default["key"] = "value"'
[2017-02-12T17:24:30+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Converge failed on instance <default-centos-6>.  Please see .kitchen/logs/default-centos-6.log for more details
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
```

### PR DESCRIPTION
**Pull request by lamont-granquist**
as noted in comments this works around bugginess in the yum
provider that require a major refactor to fix.

closes #5812

### CODE REVIEW

File: lib/chef/resource/yum_package.rb [hunk 1]
```diff
@@ -24,8 +24,17 @@ class YumPackage < Chef::Resource::Package
...
-      # Install a specific arch
-      property :arch, [ String, Array ]
+      # XXX: the coercions here are due to the provider promiscuously updating the properties on the
+      # new_resource which causes immutable modification exceptions when passed an immutable node array.
+      #
+      # <lecture>
+      # THIS is why updating the new_resource in a provider is so terrible, and is equivalent to methods scribbling over
+      # its own arguments as unintended side-effects (and why functional languages that don't allow modifcations
+      # of variables eliminate entire classes of bugs).
+      # </lecture>
```
Comment (coderanger) [inline lib/chef/resource/yum_package.rb]: maxresdefault-1

**Comment by coderanger** [comment]
Just for others the might find this in the future, `Chef::Node::ImmutableArray#to_a` does a copy internally so this isn't just a "cast away the const-ness".

**Comment by lamont-granquist** [comment]
yes i used that because i know it does an inherent `#dup`

for chef 13 we really need to freeze the string things in immutable hash/arrays and make sure that dup really dups the whole thing (because i think you'd find right now that if i mutated the strings with `<<` and not just the array elements that chef would be fine with that and the default level attr that i set would be getting updated-in-place -- which is the kind of thing which led facebook to mandate calling `#to_hash` all over the place in order to dup attributes before playing around with them...  which is probably way more than anyone ever wanted to know...)

**Comment by lamont-granquist** [comment]
@chef/client-core today's version of coding horror for your review and amusement...

**Pull request merged**